File: sqlite_database/generic_database.c
// gcc sqlite3.o generic_database.c && ./a.out
// compile sqlite by downloading an amalgamated version and doing  "gcc -c sqlite.c" to get the object file.

#include <stdio.h>
#include "sqlite3.h"

// Define maximum length for strings
#define MAX_LENGTH 50

// Callback function to handle results of queries
int callback(void *NotUsed, int argc, char **argv, char **azColName) {
    for (int i = 0; i < argc; i++) {
        printf("%s = %s\n", azColName[i], argv[i] ? argv[i] : "NULL");
    }
    printf("\n");
    return 0;
}

// Function to open or create a database
sqlite3 *open_database(const char *db_name) {
    sqlite3 *db;
    int rc = sqlite3_open(db_name, &db);
    if (rc != SQLITE_OK) {
        fprintf(stderr, "Can't open database: %s\n", sqlite3_errmsg(db));
        sqlite3_close(db);
        return NULL;
    }
    return db;
}

// Function to close a database
void close_database(sqlite3 *db) {
    sqlite3_close(db);
}

// Function to create a table
void create_table(sqlite3 *db) {
    char *sql = "CREATE TABLE IF NOT EXISTS elements("
                "id INTEGER PRIMARY KEY AUTOINCREMENT,"
                "name TEXT NOT NULL,"
                "value TEXT NOT NULL);";
    int rc = sqlite3_exec(db, sql, callback, 0, NULL);
    if (rc != SQLITE_OK) {
        fprintf(stderr, "SQL error: %s\n", sqlite3_errmsg(db));
    } else {
        printf("Table created successfully\n");
    }
}

// Function to insert a row into the table
void insert_row(sqlite3 *db, const char *name, const char *value) {
    char sql[MAX_LENGTH * 2 + 100];
    sprintf(sql, "INSERT INTO elements (name, value) VALUES ('%s', '%s');", name, value);
    int rc = sqlite3_exec(db, sql, callback, 0, NULL);
    if (rc != SQLITE_OK) {
        fprintf(stderr, "SQL error: %s\n", sqlite3_errmsg(db));
    } else {
        printf("Row inserted successfully\n");
    }
}

// Function to select all rows from the table
void select_rows(sqlite3 *db) {
    char *sql = "SELECT * FROM elements;";
    int rc = sqlite3_exec(db, sql, callback, 0, NULL);
    if (rc != SQLITE_OK) {
        fprintf(stderr, "SQL error: %s\n", sqlite3_errmsg(db));
    }
}

// Function to update a row in the table
void update_row(sqlite3 *db, int id, const char *name, const char *value) {
    char sql[MAX_LENGTH * 2 + 100];
    sprintf(sql, "UPDATE elements SET name = '%s', value = '%s' WHERE id = %d;", name, value, id);
    int rc = sqlite3_exec(db, sql, callback, 0, NULL);
    if (rc != SQLITE_OK) {
        fprintf(stderr, "SQL error: %s\n", sqlite3_errmsg(db));
    } else {
        printf("Row updated successfully\n");
    }
}

// Function to delete a row from the table
void delete_row(sqlite3 *db, int id) {
    char sql[100];
    sprintf(sql, "DELETE FROM elements WHERE id = %d;", id);
    int rc = sqlite3_exec(db, sql, callback, 0, NULL);
    if (rc != SQLITE_OK) {
        fprintf(stderr, "SQL error: %s\n", sqlite3_errmsg(db));
    } else {
        printf("Row deleted successfully\n");
    }
}

int main() {
    sqlite3 *db = open_database("example.db");
    if (db == NULL) {
        return 1;
    }

    create_table(db);

    insert_row(db, "Example 1", "Value 1");
    insert_row(db, "Example 2", "Value 2");
    select_rows(db);

    update_row(db, 1, "Updated 1", "Updated Value 1");
    delete_row(db, 2);
    select_rows(db);

    close_database(db);

    return 0;
}

File: sqlite_database/generic_database.py
import sqlite3

class GenericDatabase:
    def __init__(self, db_name='generic_db.sqlite'):
        self.db_name = db_name

    def connect_db(self):
        return sqlite3.connect(self.db_name)

    def create_table(self, table_name, columns):
        conn = self.connect_db()
        cursor = conn.cursor()
        columns_definition = ', '.join([f'{col_name} {col_type}' for col_name, col_type in columns.items()])
        cursor.execute(f'''
            CREATE TABLE IF NOT EXISTS {table_name} (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                {columns_definition}
            )
        ''')
        conn.commit()
        conn.close()

    def create_element(self, table_name, element):
        conn = self.connect_db()
        cursor = conn.cursor()
        columns = ', '.join(element.keys())
        placeholders = ', '.join(['?' for _ in element])
        cursor.execute(f'''
            INSERT INTO {table_name} ({columns})
            VALUES ({placeholders})
        ''', tuple(element.values()))
        conn.commit()
        conn.close()

    def read_elements(self, table_name):
        conn = self.connect_db()
        cursor = conn.cursor()
        cursor.execute(f'SELECT * FROM {table_name}')
        elements = cursor.fetchall()
        conn.close()
        return elements

    def update_element(self, table_name, element_id, updates):
        conn = self.connect_db()
        cursor = conn.cursor()
        set_clause = ', '.join([f'{col} = ?' for col in updates.keys()])
        cursor.execute(f'''
            UPDATE {table_name}
            SET {set_clause}
            WHERE id = ?
        ''', (*updates.values(), element_id))
        conn.commit()
        conn.close()

    def delete_element(self, table_name, element_id):
        conn = self.connect_db()
        cursor = conn.cursor()
        cursor.execute(f'''
            DELETE FROM {table_name}
            WHERE id = ?
        ''', (element_id,))
        conn.commit()
        conn.close()


if __name__ == "__main__":

    # Initialize the database
    db = GenericDatabase()

    # Define a table with its columns
    columns = {
        'name': 'TEXT NOT NULL',
        'value': 'TEXT NOT NULL'
    }
    db.create_table('elements', columns)

    # Create a new element
    db.create_element('elements', {'name': 'example_name', 'value': 'example_value'})

    # Read all elements
    elements = db.read_elements('elements')
    print(elements)

    # Update an element
    db.update_element('elements', 1, {'name': 'updated_name', 'value': 'updated_value'})

    # Delete an element
    db.delete_element('elements', 1)

    # Check elements again
    elements = db.read_elements('elements')
    print(elements)



    # Define a table with its columns
    columns = {
        'a': 'TEXT NOT NULL',
        'b': 'TEXT NOT NULL',
        'c': 'TEXT NOT NULL'        
        
    }
    table_name = 'elements2'
    
    db.create_table(table_name, columns)

    # Create a new element
    db.create_element(table_name, {'a': 'omg', 'b':'zomg', 'c':'yes omgomg'})
    db.create_element(table_name, {'a': 'omg', 'b':'zomg', 'c':'yes omgomg'})
    db.create_element(table_name, {'a': 'omg', 'b':'zomg', 'c':'yes omgomg'})
    db.create_element(table_name, {'a': 'omg', 'b':'zomg', 'c':'yes omgomg'})    

    elements = db.read_elements(table_name)
    print("Elements2")
    print(elements)

    # Update an element
    db.update_element(table_name, 1, {'a': 'updated_name', 'b': 'updated_value'})

    # Delete an element
    db.delete_element(table_name, 1)

    # Check elements again
    elements = db.read_elements(table_name)
    print(elements)
    
